claat/fetch/fetch.go

```diff
@@ -13,8 +13,419 @@
 // limitations under the License.
 package fetch
 
-import "io"
+import (
+	"encoding/json"
+	"fmt"
+	"hash/crc64"
+	"io"
+	"io/ioutil"
+	"math"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
 
-type Fetcher interface {
-	Fetch() (io.Reader, error)
+	"github.com/googlecodelabs/tools/claat/fetch/drive/auth"
+	"github.com/googlecodelabs/tools/claat/parser"
+	"github.com/googlecodelabs/tools/claat/types"
+	"github.com/googlecodelabs/tools/claat/util"
+)
+
+const (
+	// supported codelab source types must be registered parsers
+	// TODO: define these in claat/parser/..., e.g. in parser/gdoc
+	// alternate TODO: make this an iota-based enum?
+	SrcInvalid   srcType = ""
+	SrcGoogleDoc srcType = "gdoc" // Google Docs doc
+	SrcMarkdown  srcType = "md"   // Markdown text
+
+	// driveAPI is a base URL for Drive API
+	driveAPI = "https://www.googleapis.com/drive/v3"
+)
+
+// TODO: create an enum for use with "nometa" for readability's sake
+
+// srcType is codelab source type
+type srcType string
+
+// resource is a codelab resource, loaded from local file
+// or fetched from remote location.
+type resource struct {
+	typ  srcType       // source type
+	body io.ReadCloser // resource body
+	mod  time.Time     // last update of content
+}
+
+// codelab wraps types.Codelab, while adding source type
+// and modified timestamp fields.
+type codelab struct {
```

Comment:
Should this be exported since it's returned in the exported func SlurpCodelab below?

Reply:
There's not much of a reason for other code to create codelab structs. Or is this type of thing a general anti-idiom and it should be exported just for readability's sake? WDYT?

Reply:
This looks to be one of those issues with people on both sides, so this should be fine then.